### PR TITLE
[Bug] Another fix on PyTorch memory leakage issue

### LINKED ISSUE #997: GPU OOM for running GCN example code
## 🐛 Bug

I tried to follow GCN code described in https://docs.dgl.ai/en/latest/tutorials/models/1_gnn/1_gcn.html#sphx-glr-tutorials-models-1-gnn-1-gcn-py

When I run the training snippet after adding one line, OOM occurs.

```python
import time
import numpy as np
g, features, labels, mask = load_cora_data()
optimizer = th.optim.Adam(net.parameters(), lr=1e-3)
dur = []
for epoch in range(30):
    if epoch >=3:
        t0 = time.time()

    logits = net(g, features)
    logp = F.log_softmax(logits, 1)
    loss = F.nll_loss(logp[mask], labels[mask])

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    if epoch >=3:
        dur.append(time.time() - t0)

    print("Epoch {:05d} | Loss {:.4f} | Time(s) {:.4f}".format(
            epoch, loss.item(), np.mean(dur)))

    logits = net(g, features)    #### Added one line

```
## To Reproduce

1. Add `logits = net(g, features)` in the training loop.
1. Run the training snippet.
1. 

## Expected behavior
`logits = net(g, features)` is just a forward computation and it should not cause OOM.
I guess that DGL internally reallocates memory for the same parameters.
This one line is useful to check `validation error` which is used to know when overfitting happens.

## Environment

 - DGL Version (e.g., 1.0): 0.4
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): PyTorch 1.2.0
 - OS (e.g., Linux): Linux
 - How you installed DGL (`conda`, `pip`, source): `conda`
 - Build command you used (if compiling from source): 
 - Python version: 3.6.9
 - CUDA/cuDNN version (if applicable): CUDA 9.2
 - GPU models and configuration (e.g. V100): GTX 1080 Ti
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
According to https://github.com/pytorch/pytorch/issues/25340, apparently input tensors should be `save_for_backward`'ed as well.  Credit for @yzh119 for finding this.

This should also fix #997 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by BarclayII** [comment]
Apparently GAT example still suffers from this.

**Comment by BarclayII** [comment]
Rewriting EdgeSoftmax module with `F.binary_reduce` and `F.copy_reduce` solves the leakage problem in the GAT example.  @jermainewang any ideas?

**Comment by yzh119** [comment]
@BarclayII Note that we cached `graph_index` in https://github.com/dmlc/dgl/blob/e890a899fc31bb27be848a60e1f20715e9a3cb1e/python/dgl/graph_index.py#L664.

Could this be a possible reason for memory leak?

**Comment by yzh119** [comment]
However, OOM still exists after I removed the decorator...

**Comment by yzh119** [comment]
You are sexy

**Comment by BarclayII** [comment]
@yzh119 Did you check if this fix work in your setting?

Also I wonder if sexy is a correct word here.
